crates/polars-core/src/chunked_array/arithmetic/decimal.rs

```diff
@@ -1,119 +1,44 @@
-                let opt_lhs = lhs.get(0);
-                match opt_lhs {
-                    None => ChunkedArray::full_null(lhs.name(), rhs.len()),
-                    Some(lhs_val) => {
-                        let chunks = rhs
-                            .downcast_iter()
-                            .map(|rhs| operation_rhs(lhs_val, rhs).map(|a| Box::new(a) as ArrayRef))
-                            .collect::<PolarsResult<_>>()?;
-                        unsafe { lhs.copy_with_chunks(chunks, false, false) }
-                    },
-                }
-            },
-            _ => {
-                polars_bail!(ComputeError: "Cannot apply operation on arrays of different lengths")
-            },
-        };
-        ca.rename(lhs.name());
-        Ok(ca.into_decimal_unchecked(self.precision(), self.scale()))
-    }
-}
 
 impl Add for &DecimalChunked {
     type Output = PolarsResult<DecimalChunked>;
 
     fn add(self, rhs: Self) -> Self::Output {
-        self.arithmetic_helper(
-            rhs,
-            decimal::add,
-            |lhs, rhs_val| decimal::add_scalar(lhs, rhs_val, &rhs.dtype().to_arrow(true)),
-            |lhs_val, rhs| decimal::add_scalar(rhs, lhs_val, &self.dtype().to_arrow(true)),
-        )
+        let scale = self.scale().max(rhs.scale());
+        let lhs = self.to_scale(scale)?;
+        let rhs = rhs.to_scale(scale)?;
+        Ok((&lhs.0 + &rhs.0).into_decimal_unchecked(None, scale))
     }
 }
 
 impl Sub for &DecimalChunked {
     type Output = PolarsResult<DecimalChunked>;
 
     fn sub(self, rhs: Self) -> Self::Output {
-        self.arithmetic_helper(
-            rhs,
-            decimal::sub,
-            decimal::sub_scalar,
-            decimal::sub_scalar_swapped,
-        )
+        let scale = self.scale().max(rhs.scale());
+        let lhs = self.to_scale(scale)?;
+        let rhs = rhs.to_scale(scale)?;
+        Ok((&lhs.0 - &rhs.0).into_decimal_unchecked(None, scale))
     }
 }
 
 impl Mul for &DecimalChunked {
     type Output = PolarsResult<DecimalChunked>;
 
     fn mul(self, rhs: Self) -> Self::Output {
-        self.arithmetic_helper(
-            rhs,
-            decimal::mul,
-            |lhs, rhs_val| decimal::mul_scalar(lhs, rhs_val, &rhs.dtype().to_arrow(true)),
-            |lhs_val, rhs| decimal::mul_scalar(rhs, lhs_val, &self.dtype().to_arrow(true)),
-        )
+        let scale = self.scale() + rhs.scale();
+        Ok((&self.0 * &rhs.0).into_decimal_unchecked(None, scale))
     }
 }
 
 impl Div for &DecimalChunked {
     type Output = PolarsResult<DecimalChunked>;
 
     fn div(self, rhs: Self) -> Self::Output {
-        self.arithmetic_helper(
-            rhs,
-            decimal::div,
-            |lhs, rhs_val| decimal::div_scalar(lhs, rhs_val, &rhs.dtype().to_arrow(true)),
-            |lhs_val, rhs| decimal::div_scalar_swapped(lhs_val, &self.dtype().to_arrow(true), rhs),
-        )
+        // Follow postgres and MySQL adding a fixed scale increment of 4
```

Comment:
Interesting. And then they accept that after a few multiplication it simply doesn't fit?

Reply:
Actually, the comment is from https://github.com/apache/arrow-rs/blob/0dda129be715a95981f0151c8378f0b0b37a4ae9/arrow-arith/src/numeric.rs#L791